File: ros/src/lidar_pkg/src/lidar_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <geometry_msgs/Twist.h>

ros::Publisher vel_pub;
int count = 0;

void LidarCallback(const sensor_msgs::LaserScan msg)
{
   float fMidDist = msg.ranges[180];
   ROS_INFO("前方测距 ranges[180] = %f 米", fMidDist);

   if(count > 0)
   {
      count--;
      return;
   }

   geometry_msgs::Twist vel_cmd;
   if(fMidDist < 1.5)
   {
      vel_cmd.angular.z = 0.3;
      count = 50;
   }else
   {
      vel_cmd.linear.x = 0.05;
   }

   vel_pub.publish(vel_cmd);

}

 int main(int argc, char *argv[])
 {
   setlocale(LC_ALL, "");
   ros::init(argc, argv, "lidar_node");

   ros::NodeHandle n;
   ros::Subscriber lidar_sub = n.subscribe("/scan", 10, &LidarCallback);
   vel_pub = n.advertise<geometry_msgs::Twist>("/cmd_vel", 10);


   ros::spin();


   return 0;
 }
 

File: ros/src/lidar_pkg/scripts/lidar_node.py
#!/usr/bin/env python3
#coding=utf-8

import rospy
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist

count = 0


def lidarCallback(msg):
    global vel_pub
    global count
    dist = msg.ranges[180]
    rospy.loginfo("前方测距 range[180] = %f 米" ,dist)

    if count > 0:
        count -= 1
        return
    vel_cmd = Twist()
    if dist < 1.5:
        count = 50
        vel_cmd.angular.z = 0.3
    else:
        vel_cmd.linear.x = 0.1
    vel_pub.publish(vel_cmd)
    



if __name__ == "__main__":
    rospy.init_node("lidar_node")

    lidar_sub = rospy.Subscriber("/scan", LaserScan, lidarCallback, queue_size=10)
    vel_pub = rospy.Publisher("cmd_vel", Twist, queue_size=10)
    rospy.spin()